marconi/src/Marconi/Index/ScriptTx.hs

```diff
@@ -32,110 +31,265 @@ import Cardano.Ledger.Keys qualified as LedgerShelley
+            , SQL.toField $ txSlot o ]
 
-type Query = ScriptAddress
-type Result = [TxCbor]
-
-data ScriptTxUpdate = ScriptTxUpdate
-  { txScripts :: [(TxCbor, [ScriptAddress])]
-  , slotNo    :: !SlotNo
-  } deriving (Show)
+deriving anyclass instance SQL.FromRow ScriptTxRow
 
-type ScriptTxIndex = SqliteIndex ScriptTxUpdate () Query Result
+-- * Indexer
 
+type Query = StorableQuery ScriptTxHandle
+type Result = StorableResult ScriptTxHandle
 
-toUpdate :: forall era . C.IsCardanoEra era => [C.Tx era] -> SlotNo -> ScriptTxUpdate
-toUpdate txs = ScriptTxUpdate txScripts'
+toUpdate
+  :: forall era . C.IsCardanoEra era
+  => [C.Tx era]
+  -> SlotNo
+  -> StorableEvent ScriptTxHandle
+toUpdate txs = ScriptTxEvent txScripts'
   where
     txScripts' = map (\tx -> (TxCbor $ C.serialiseToCBOR tx, getTxScripts tx)) txs
 
-getTxBodyScripts :: forall era . C.TxBody era -> [ScriptAddress]
+getTxBodyScripts :: forall era . C.TxBody era -> [StorableQuery ScriptTxHandle]
 getTxBodyScripts body = let
     hashesMaybe :: [Maybe C.ScriptHash]
     hashesMaybe = case body of
-      Shelley.ShelleyTxBody shelleyBasedEra _ scripts _ _ _ -> flip map scripts $ \script ->
-        case fromShelleyBasedScript shelleyBasedEra script of
-          Shelley.ScriptInEra _ script' -> Just $ C.hashScript script'
+      Shelley.ShelleyTxBody shelleyBasedEra _ scripts _ _ _ ->
+        flip map scripts $ \script ->
+          case fromShelleyBasedScript shelleyBasedEra script of
+            Shelley.ScriptInEra _ script' -> Just $ C.hashScript script'
       _ -> [] -- Byron transactions have no scripts
     hashes = catMaybes hashesMaybe :: [Shelley.ScriptHash]
-  in map ScriptAddress hashes
+  in map ScriptTxAddress hashes
 
-getTxScripts :: forall era . C.Tx era -> [ScriptAddress]
+getTxScripts :: forall era . C.Tx era -> [StorableQuery ScriptTxHandle]
 getTxScripts (C.Tx txBody _ws) = getTxBodyScripts txBody
 
-open :: (ScriptTxIndex -> ScriptTxUpdate -> IO [()]) -> FilePath -> Depth -> IO ScriptTxIndex
-open onInsert dbPath (Depth k) = do
-  ix <- fromJust <$> Ix.newBoxed query store onInsert k ((k + 1) * 2) dbPath
-  let c = ix ^. Ix.handle
-  SQL.execute_ c "CREATE TABLE IF NOT EXISTS script_transactions (scriptAddress TEXT NOT NULL, txCbor BLOB NOT NULL)"
-  SQL.execute_ c "CREATE INDEX IF NOT EXISTS script_address ON script_transactions (scriptAddress)"
-  pure ix
+{- Now that all connected data types have been defined, we go on to implement some
+   of the type classes required for information storage and retrieval. -}
+
+instance Buffered ScriptTxHandle where
+  {- The data is buffered in memory. When the memory buffer is filled, we need to store
+     it on disk. -}
+  persistToStorage
+    :: Foldable f
+    => f (StorableEvent ScriptTxHandle)
+    -> ScriptTxHandle
+    -> IO ScriptTxHandle
+  persistToStorage es h = do
+    let rows = foldl' (\ea e -> ea ++ flatten e) [] es
+        c    = hdlConnection h
+    SQL.execute_ c "BEGIN"
+    forM_ rows $
```

Comment:
nit-pic
use [executeMany](https://hackage.haskell.org/package/sqlite-simple-0.4.18.2/docs/Database-SQLite-Simple.html#v:executeMany) to take advantage of SQL batching?